pkg/ddl/ddl_running_jobs.go

```diff
@@ -19,119 +19,158 @@
 package ddl
 
 import (
+	"fmt"
 	"strconv"
 	"strings"
 	"sync"
 
 	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/util/intest"
 )
 
 type runningJobs struct {
-	sync.RWMutex
-	// processingIDs records the IDs of the jobs that are being processed by a worker.
-	processingIDs    map[int64]struct{}
-	processingIDsStr string
-
-	// unfinishedIDs records the IDs of the jobs that are not finished yet.
-	// It is not necessarily being processed by a worker.
-	unfinishedIDs    map[int64]struct{}
-	unfinishedSchema map[string]map[string]struct{} // database -> table -> struct{}
+	// although most of the usage is called by jobScheduler in a single goroutine,
+	// runningJobs.remove is called by the worker goroutine. Another implementation
+	// is let worker goroutine send the finished job to jobScheduler, and
+	// jobScheduler calls remove, so no need to lock.
+	mu sync.RWMutex
+
+	ids          map[int64]struct{}
+	idsStrGetter func() string
+	// database -> table -> struct{}
+	//
+	// if the job is only related to a database, the table-level entry key is
+	// model.InvolvingAll. When remove a job, runningJobs will make sure no
+	// zero-length map exists in table-level.
+	schemas map[string]map[string]struct{}
 }
 
 func newRunningJobs() *runningJobs {
 	return &runningJobs{
-		processingIDs:    make(map[int64]struct{}),
-		unfinishedSchema: make(map[string]map[string]struct{}),
-		unfinishedIDs:    make(map[int64]struct{}),
+		ids:          make(map[int64]struct{}),
+		idsStrGetter: func() string { return "" },
+		schemas:      make(map[string]map[string]struct{}),
 	}
 }
 
-func (j *runningJobs) clear() {
-	j.Lock()
-	defer j.Unlock()
-	j.unfinishedIDs = make(map[int64]struct{})
-	j.unfinishedSchema = make(map[string]map[string]struct{})
-}
+// add should only add the argument that passed the last checkRunnable.
+func (j *runningJobs) add(jobID int64, involves []model.InvolvingSchemaInfo) {
+	j.mu.Lock()
+	defer j.mu.Unlock()
 
-func (j *runningJobs) add(jobID int64, involvedSchemaInfos []model.InvolvingSchemaInfo) {
-	j.Lock()
-	defer j.Unlock()
-	j.processingIDs[jobID] = struct{}{}
-	j.updateInternalRunningJobIDs()
+	j.ids[jobID] = struct{}{}
+	j.updateIDsStrGetter()
 
-	if _, ok := j.unfinishedIDs[jobID]; ok {
-		// Already exists, no need to add it again.
-		return
-	}
-	j.unfinishedIDs[jobID] = struct{}{}
-	for _, info := range involvedSchemaInfos {
-		if _, ok := j.unfinishedSchema[info.Database]; !ok {
-			j.unfinishedSchema[info.Database] = make(map[string]struct{})
+	for _, info := range involves {
+		if info.Database == model.InvolvingNone {
+			// should not happen
+			if intest.InTest {
+				if info.Table != model.InvolvingNone {
+					panic(fmt.Sprintf(
+						"job %d is invalid, involved table name is not empty: %s",
+						jobID, info.Table,
+					))
+				}
+			}
```

Comment:
There are some invalid jobs in tests(like `doDDLJobErr`). Maybe you can search `&model.Job{` and fix all usages.

Reply:
I only check for db is None but table is not None. Not found in CI I think.